modules/stochastic_tools/src/utils/CovarianceFunction.C

```diff
@@ -0,0 +1,402 @@
+  {
+    dataStore(stream, _length_factor.at(ii), context);
+  }
+}
+// end squared_exponential
+
+// begin exponential
+Exponential::Exponential() : CovarianceKernel() {}
+
+RealEigenMatrix
+Exponential::compute_K(const RealEigenMatrix x,
+                       const RealEigenMatrix xp,
+                       const bool is_self_covariance) const
+{
+  unsigned int num_samples_x = x.rows();
+  unsigned int num_samples_xp = xp.rows();
+  unsigned int num_params_x = x.cols();
+  unsigned int num_params_xp = xp.cols();
+
+  if (num_params_x != num_params_xp)
+  {
+    ::mooseError("Number of parameters do not match in covariance kernel calculation");
+  }
+
+  RealEigenMatrix K(num_samples_x, num_samples_xp);
+
+  for (unsigned int ii = 0; ii < num_samples_x; ii++)
+  {
+    for (unsigned int jj = 0; jj < num_samples_xp; jj++)
+    {
+      Real r_scaled = 0;
+      // Compute distance per parameter, scaled by length factor
+      for (unsigned int kk = 0; kk < num_params_x; kk++)
+      {
+        r_scaled += pow((x(ii, kk) - xp(jj, kk)) / _length_factor.at(kk), 2);
+      }
+      r_scaled = sqrt(r_scaled);
+      K(ii, jj) = _sigma_f_squared * std::exp(-pow(r_scaled, _gamma));
+    }
+    if (is_self_covariance)
+    {
+      K(ii, ii) += _sigma_n_squared;
+    }
+  }
+
+  return K;
+}
+
+void
+Exponential::store(std::ostream & stream, void * context) const
+{
+  std::string type = "Exponential";
+  dataStore(stream, type, context);
+  dataStore(stream, _sigma_f_squared, context);
+  dataStore(stream, _sigma_n_squared, context);
+  unsigned int n = _length_factor.size();
+  dataStore(stream, n, context);
+  for (unsigned int ii = 0; ii < n; ii++)
+  {
+    dataStore(stream, _length_factor.at(ii), context);
+  }
+  dataStore(stream, _gamma, context);
+}
+
+void
+Exponential::set_gamma(const Real gamma)
+{
+  _gamma = gamma;
+}
+// end exponential
+
+// begin matern half integer
+MaternHalfInt::MaternHalfInt() : CovarianceKernel() {}
+
+RealEigenMatrix
+MaternHalfInt::compute_K(const RealEigenMatrix x,
+                         const RealEigenMatrix xp,
+                         const bool is_self_covariance) const
+{
+  unsigned int num_samples_x = x.rows();
+  unsigned int num_samples_xp = xp.rows();
+  unsigned int num_params_x = x.cols();
+  unsigned int num_params_xp = xp.cols();
+
+  if (num_params_x != num_params_xp)
+  {
+    ::mooseError("Number of parameters do not match in covariance kernel calculation");
+  }
+
+  RealEigenMatrix K(num_samples_x, num_samples_xp);
+
+  // This factor is used over and over, don't calculate each time
+  Real factor = sqrt(2 * _p + 1);
+
+  for (unsigned int ii = 0; ii < num_samples_x; ii++)
+  {
+    for (unsigned int jj = 0; jj < num_samples_xp; jj++)
+    {
+      Real r_scaled = 0;
+      // Compute distance per parameter, scaled by length factor
+      for (unsigned int kk = 0; kk < num_params_x; kk++)
+      {
+        r_scaled += pow((x(ii, kk) - xp(jj, kk)) / _length_factor.at(kk), 2);
+      }
```

Comment:
If it is only one line in the for loop, then you can get rid of the braces. This makes things look nicer.